firmware-management/mbl-watchdog-init/mbl-watchdog-init.c

```diff
@@ -0,0 +1,173 @@
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *
+ * This utility initialises the hardware watchdog and sets the timeout.
+ * The program also tries to determine the last boot reason so we can tell if
+ * the last reboot was caused by the hardware watchdog.
+ * This code is intended for use with our robust update mechanism; we need
+ * to start the watchdog before we mount the rootfs, so the device reboots if
+ * loading the rootfs fails.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <linux/watchdog.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+//============================================================================
+// Error handling and logging convenience functions
+int is_err(const int ret)
+{
+    return ret == -1;
+}
+
+
+int log_error(const char *message)
+{
+    int err_code = errno;
+    char *prefix = "WATCHDOG-ERROR:";
+    char result[strlen(prefix) + strlen(message)];
+
+    snprintf(result, sizeof(result), "%s %s", prefix, message);
+    perror(result);
+
+    return err_code;
+}
+
+
+void log_warning(const char *message)
+{
+    char *prefix = "WATCHDOG-WARNING:";
+    printf("%s %s\n", prefix, message);
+}
+
+
+void log_info(const char *message)
+{
+    char *prefix = "WATCHDOG-INFO:";
+    printf("%s %s\n", prefix, message);
+}
+
+//============================================================================
+// Watchdog functions
+int set_watchdog_timeout(const int watchdog_fd, int *timeout)
+{
+    return ioctl(watchdog_fd, WDIOC_SETTIMEOUT, timeout);
+}
+
+
+int print_last_boot_reason(const int boot_status)
+{
+    switch (boot_status)
+    {
+        case WDIOF_OVERHEAT:
+            log_info("The last reboot was caused by the CPU overheating.");
+            return 0;
+        case WDIOF_CARDRESET:
+            log_info("The last reboot was caused by a watchdog reset.");
+            return 0;
+        case WDIOF_FANFAULT:
+            log_info("The last reboot was because a system fan monitored by the watchdog card failed.");
+            return 0;
+        case WDIOF_EXTERN1:
+            log_info("The last reboot was because external monitoring relay/source 1 was triggered.");
+            return 0;
+        case WDIOF_EXTERN2:
+            log_info("The last reboot was because external monitoring relay/source 2 was triggered.");
+            return 0;
+        case WDIOF_POWERUNDER:
+            log_info("The last reboot was due to the machine showing an undervoltage status.");
+            return 0;
+        case WDIOF_POWEROVER:
+            log_info("The last reboot was due to the machine showing an overvoltage status.");
+            return 0;
+    }
+
+    return -1;
+}
+
+
+int get_last_boot_status(const int watchdog_fd)
+{
+    int flags;
+
+    const int gs_ret = ioctl(watchdog_fd, WDIOC_GETBOOTSTATUS, &flags);
+    if (is_err(gs_ret))
+    {
+        return -1;
```

Comment:
Why you are not returning `gs_ret`? Returning and printing the error code from `WDIOC_GETBOOTSTATUS` can be useful for debugging and bug fixing.

Reply:
IIUC `gs_ret` is always -1 when an error occurs, the error code is stored in `errno` which is returned from the `log_error` function.

Reply:
cb306fb